[c10] add an option to pg_config split share

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130877

Summary:
context is: #129865
We want to give users an option to not share comms resouces so that
comm opts can overlap
Test Plan:
Augmentd UT

Tags:

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

shuqiangzhang committed:
[c10] add an option to pg_config split share
Summary:
context is: #129865
We want to give users an option to not share comms resouces so that
comm opts can overlap
Test Plan:
Augmentd UT

Tags:

[ghstack-poisoned]

shuqiangzhang commented:
cc @nvcastet

shuqiangzhang committed:
Update on "[c10] add an option to pg_config split share"


Summary:
context is: #129865
We want to give users an option to not share comms resouces so that
comm opts can overlap
Test Plan:
Augmentd UT

Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

nvcastet commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2784,6 +2785,7 @@ Example::
     >>> nccl_options.config.cga_cluster_size = 2
     >>> nccl_options.config.max_ctas = 4
     >>> nccl_options.config.min_ctas = 2
+    >>> nccl_options.config.split_share = 1
```
I would put `split_share = 0` for the example here since you plan to put the default to `1`.

shuqiangzhang replied:
In this version of code, split_share = 0 is the default

shuqiangzhang commented:
> > cc @nvcastet, cannot init config.splitShare = 1 after `ncclConfig_t config = NCCL_CONFIG_INITIALIZER;' due to compile time error: ‘config’ does not name a type
> 
> Yes you need to put it inside the `#ifdef` since `config` is only declared there. Inside that `#ifdef` you would also need to add a nested `#ifdef NCCL_HAS_COMM_SPLIT` to make sure `splitShare` is present in that NCCL version.

I did exactly both yesterday, the compiler was still not happy

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2796,6 +2796,7 @@ def test_pass_nccl_options_config(self):
         pg_opts.config.min_ctas = 2
         pg_opts.config.cga_cluster_size = 2
         pg_opts.config.net_name = "Socket"
+        pg_opts.config.split_share = 1
```
Not necessary to do it in this PR. Do you think we want to add some docs in distributed_c10d.py about NCCL config in the option? Currently the only info we have is by reading through cpp code and unit test, IIUC.

shuqiangzhang replied:
more clearer doc is always welcome ^_^

wconstab replied:
but at least we should put a comment saying what this does?  i'm actually not clear on when i would change it-

is it this? or something else?
1= share resources so less memory and sockets are used (unclear how much savings here vs the tradeoff of 0)
0= can overlap p2p comms but waste more communicators and init time?

shuqiangzhang replied:
All those are NCCL specific configs, which are documented by NCCL as the source of truth and referred in the Pybind file (init.cpp) line 2739: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/types.html#ncclconfig-t. 
<img width="831" alt="image" src="https://github.com/user-attachments/assets/17e73179-835a-41a3-8af2-8a0f734f9f93">. Specifically, for this config, it explains as "Specify whether to share resources with child communicator during communicator split. Set the value of splitShare to 0 or 1. The default value is 0. When the parent communicator is created with splitShare=1 during ncclCommInitRankConfig, the child communicator can share internal resources of the parent during communicator split. Split communicators are in the same family. When resources are shared, aborting any communicator can result in other communicators in the same family becoming unusable. Irrespective of whether sharing resources or not, users should always abort/destroy all no longer needed communicators to free up resources.

"

shuqiangzhang commented:
cc @mengph , if this PR is accepted as is (splitShare=0) would be the default consistent with NCCL, so you will have to set the pg_opts.config.split_share = 1 and pass it to init_process_group if you want the 'share' behavior

fduwjj reviewed (approved):
Stamped now and as we have discussed offline, it will be great if we can surface the doc for the nccl config a little bit more.

shuqiangzhang committed:
Update on "[c10] add an option to pg_config split share"


Summary:
context is: #129865
We want to give users an option to not share comms resouces so that
comm opts can overlap
Test Plan:
Augmentd UT

Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge -f "unrelated lint warnings"

atalman commented:
@pytorchmergebot revert -c ghfirst -m "breaks internal build"

shuqiangzhang committed:
Update on "[c10] add an option to pg_config split share"


Summary:
context is: #129865
We want to give users an option to not share comms resouces so that
comm opts can overlap
Test Plan:
Augmentd UT

Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge -f "another try"
